rep-2017.rst

```diff
@@ -0,0 +1,359 @@
+  Brief:
+    Free the list of thread attributes.
+  Parameters:
+    [in] thread_attrs    The structure to be deallocated.
+  Return value:
+    RCUTILS_RET_OK
+                if the memory was successfully freed, or
+    RCUTILS_RET_INVALID_ARGUMENT
+                if any function arguments are invalid.
+
+.. code-block:: C++
+
+  rcutils_ret_t rcutils_thread_attrs_copy(rcutils_thread_attrs_t const * thread_attrs,
+                    rcutils_thread_attrs_t * out_thread_attrs);
+  Brief:
+    Copies the list of thread attributes.
+  Parameters:
+    [in]  thread_attrs      The source list of thread attributes.
+    [out] out_thread_attrs  The destination location.
+  Return value:
+    RCUTILS_RET_OK
+                if the source list was successfully copied to the destination, or
+    RCUTILS_RET_INVALID_ARGUMENT
+                if function arguments are invalid, or
+    RCUTILS_RET_BAD_ALLOC
+                if allocating memory failed.
+
+.. code-block:: C++
+
+  rcutils_ret_t rcutils_thread_attrs_add_attr(rcutils_thread_attrs_t * thread_attrs,
+                    rcutils_thread_scheduling_policy_t sched_policy,
+                    int core_affinity,
+                    int priority,
+                    char const * name);
+  Brief:
+    Adds a thread attribute to the list of thread attributes.
+  Parameters:
+    [inout] thread_attrs    The list of thread attributes to add a thread attribute to.
+    [in]    sched_policy    The thread scheduling policy of the adding attribute.
+    [in]    core_affinity   The thread core affinity of the adding attribute.
+    [in]    priority        The thread priority of the adding attribute.
+    [in]    name            The thread name of the adding attribute.
+  Return value:
+    RCUTILS_RET_OK  if the thread attribute was successfully added, or
+    RCUTILS_RET_INVALID_ARGUMENT
+                    if any function arguments are invalid, or
+    RCUTILS_RET_BAD_ALLOC
+                    if allocating memory failed, or
+    RCUTILS_RET_ERROR
+                    an unspecified error occured.
+
+Rationale
+=========
+
+Not using ROS parameter infrastructure
+--------------------------------------
+
+According to the document about the parameter [#REF-3]_, it is associated with nodes, not processes.  Events associated with each node are executed by a single executor belonging to the process.  As a result, ROS parameters are not suitable for setting the thread attributes used in the thread pool that executes the node's events altogether.
+
+Being implemented in the rcl, not in language bindings
+------------------------------------------------------
+
+According to the document about the Client libraries [#REF-4]_, every language bindings have their own thread model.  But, the thread attributes are OS-specific, not language.  So, treating the thread attributes in the rcl commonly used by the language bindings is natural and effort-saving.
+
+Backward Compatibility
```

Comment:
what happens with ROS 2 code where people already set thread attributes using available OS functions?

Reply:
> what happens with ROS 2 code where people already set thread attributes using available OS functions?

There is no effects to the existing  ROS 2 applications because the rcl with the modification of this proposal does not affect the upper layers, including applications.
Of course, you can modify those applications to use the proposed infrastructure so that you would be able to change thread attribute parameters without modification of the code.